libsrc/pce/crt0.s

```diff
@@ -0,0 +1,178 @@
+; based on code by Ullrich von Bassewitz <uz@cc65.org>
+;
+; This must be the *first* file on the linker command line
+;
+
+        .export         _exit
+        .export         __STARTUP__ : absolute = 1      ; Mark as startup
+
+        .import         initlib, donelib
+        .import         push0, _main, zerobss
+        .import         initheap
+        .import         IRQStub
+
+        ; Linker generated
+        .import         __RAM_START__, __RAM_SIZE__
+        .import         __ROM0_START__, __ROM0_SIZE__
+        .import         __ROM_START__, __ROM_SIZE__
+        .import         __STARTUP_LOAD__,__STARTUP_RUN__, __STARTUP_SIZE__
+        .import         __CODE_LOAD__,__CODE_RUN__, __CODE_SIZE__
+        .import         __RODATA_LOAD__,__RODATA_RUN__, __RODATA_SIZE__
+        .import         __DATA_LOAD__,__DATA_RUN__, __DATA_SIZE__
+        .import         __BSS_SIZE__
+
+        .include        "pce.inc"
+        .include        "extzp.inc"
+
+        .importzp       sp
+        .importzp       ptr1,ptr2
+        .importzp       tmp1,tmp2,tmp3
+
+; ------------------------------------------------------------------------
+; Place the startup code in a special segment.
+
+        .segment "STARTUP"
+
+start:
+
+        ; setup the CPU and System-IRQ
+
+        ; Initialize CPU
+
+        sei
+        nop
+        csh                     ; set high speed CPU mode
+        nop
+        cld
+        nop
+
+        ; Setup stack and memory mapping
+        ldx     #$FF            ; Stack top ($21FF)
+        txs
+
+        ; at startup all MPRs are set to 0, so init them
+        lda     #$ff
+        tam     #%00000001      ; 0000-1FFF = Hardware page
+        lda     #$F8
+        tam     #%00000010      ; 2000-3FFF = Work RAM
+
+        ; FIXME: setup a larger block of memory to use with C-code
+        ;lda     #$F7
+        ;tam     #%00000100      ; 4000-5FFF = Save RAM
+        ;lda     #1
+        ;tam     #%00001000      ; 6000-7FFF  Page 2
+        ;lda     #2
+        ;tam     #%00010000      ; 8000-9FFF  Page 3
+        ;lda     #3
+        ;tam     #%00100000      ; A000-BFFF  Page 4
+        ;lda     #4
+        ;tam     #%01000000      ; C000-DFFF  Page 5
+        ;lda     #0
+        ;tam     #%10000000      ; e000-fFFF  hucard/syscard bank 0
+
+        ; Clear work RAM (2000-3FFF)
+        stz     <$00
+        tii     $2000, $2001, $1FFF
+
+        ; Initialize hardware
+        stz     TIMER_CTRL      ; Timer off
+        lda     #$07
+        sta     IRQ_MASK        ; Interrupts off
+        stz     IRQ_STATUS      ; Acknowledge timer
+
+        ; FIXME; i dont know why the heck this one doesnt work when called from a constructor :/
+        .import vdc_init
+        jsr     vdc_init
+
+        ; Turn on background and VD interrupt/IRQ1
+        lda     #$05
+        sta     IRQ_MASK        ; IRQ1=on
+
+        ; Clear the BSS data
+        jsr     zerobss
+
+        ; Copy the .data segment to RAM
```

Comment:
> Indeed, but copydata is only used in crt0, so you don't want jsr/rts overhead.

* The same holds true for i.e. zreobss and the constructor/destructor stuff. Not inlining them has to do with code maintainability and code reuse.
* The PCE target uses several init_* functions that could be inlined without reducing code reuse.

Maybe the drawbacks could be avoided be replacing the functions in question with macros. Maybe you want to create a pull request re-engineering this area.

Anyhow I'd still like to the see the inline code below replaced with a function call (or the CPU intrinsic pointed out by Greg).